autotest.pm

```diff
@@ -345,6 +350,49 @@ sub loadtestdir {
     }
 }
 
+sub parse_serial_output {
+    if ($bmwqemu::vars{BACKEND} eq 'qemu') {
+        parse_serial_output_qemu(@_);
+    }
+}
+
+sub parse_serial_output_qemu {
+    my ($serial_file_pos, $t) = @_;
+    # serial failures defined in distri (test can override them)
+    my $failures = $t->{serial_failures};
+
+    myjsonrpc::send_json($isotovideo, {cmd => 'read_serial', position => $serial_file_pos});
+    my $json = myjsonrpc::read_json($isotovideo);
+
+    my $die = 0;
+    # loop line by line
+    for my $line (split(/^/, $json->{serial})) {
+        chomp $line;
+        # only two keys allowed
+        for my $type (qw(soft hard)) {
+            for my $regexp (@{$failures->{$type}}) {
```

Comment:
I guess you can get rid of the double loop (and the double nesting) by looping like this:
 ```perl
for(@{$failures->{hard}}, @{$failures->{soft}}) {
....
 }
``` 

or alternatively just using hash slices if you want to group even further.

Reply:
How would i then identify the type of failure, required for the next IF ? I would then to store more info, like the ammount of exceptions in each type, before starting looping... To be honest i don't think it's worth it as it would make the algorithm confusing.